### PR TITLE
kvclient: correctly handle node shutdown

### LINKED ISSUE #113125: ccl/backupccl: TestBackupWorkerFailure failed
ccl/backupccl.TestBackupWorkerFailure [failed](https://tanzanite.cluster.engflow.com/invocation/4554492b-d778-4517-93b5-6da66c531a9f) on master @ [2205334339583cea57db33e3c0d992f672e660e8](https://github.com/cockroachdb/cockroach/commits/2205334339583cea57db33e3c0d992f672e660e8):

```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestBackupWorkerFailure3861981193
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:124: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:8892: pq: could not reload backup manifest when retrying: reading backup checkpoint: recv msg error: grpc: open /var/lib/engflow/worker/work/2/exec/_tmp/d8ba48f76500d7ecd0a36e7fe2534e5e/TestBackupWorkerFailure870532821/worker-failure/2023/10/26-072028.82/progress/BACKUP-CHECKPOINT-13cdcfcdccd2cecfd2cdc9dfcfc8c5cdcfc5cdc6d1cfc6c6cfcdcdcccbc7dfd4cfcfcfcfdfaaabbcdf92c2d4cec9d1cdcecac9cec7cac7cdfffe864317050.tmp: no such file or directory [code 2/Unknown]
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestBackupWorkerFailure3861981193
--- FAIL: TestBackupWorkerFailure (7.61s)
```
<p>Parameters: <code>attempt=1</code>
, <code>run=9</code>
, <code>shard=6</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-32757

### PR DESCRIPTION
**Pull request by andrewbaptist**
Previously when the stopper was called for DistSender it could exit the retry loop early and return a retriable error to senders. There was a check to validate this in the case of no error returned, but this didn't handle retriable errors. The new check will instead return the correct error.

Fixes: #113125
Epic: none

Release note: None

### CODE REVIEW

**Review by arulajmani — COMMENTED** [review]
Given it's late in the release cycle and how subtle some of this code error swallowing can be, this patch might benefit from @nvanbenschoten's eyes as well. 

> Reviewed 6 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-:-Nm1Ncxb1ZNrUSRV3ram:bw6air0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2027 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1Kd7L98zR1LV0iTkD:-Nm1Kd7L98zR1LV0iTkE:bn0313l) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/kvclient/kvcoord/dist_sender.go#L2027)):*
> ```Go
> 
> 	// Propagate error if either the retry closer or context done channels were
> 	// closed. This replaces the return error since when the context is closed
> ```

In cases where `pErr != nil`, should we be replacing the error or wrapping it? 
___
*[`pkg/kv/txn_external_test.go` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1IfwxDstr50Jp2I4y:-Nm1IfwxDstr50Jp2I4z:b-olnph6) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/txn_external_test.go#L258)):*
> ```Go
> 			cancelCommit()
> 			commitErr := <-commitCh
> 			require.ErrorContains(t, commitErr, "context canceled")
> ```

Is this change okay? Don't we want to return an `AmbiguousResultError` in this case based on the commentary above?

**Review by andrewbaptist — COMMENTED** [review]
I agree this is late in the cycle, and it is also an option to delay backporting this until we have seen it run on `master` for a couple weeks and nothing has come out.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-:-Nm1UVcT8ogrBlsMOqV1:bjvjad4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2027 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1Kd7L98zR1LV0iTkD:-Nm1UczSD7HK73VlK9a4:bhh9na7) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/kvclient/kvcoord/dist_sender.go#L2027)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

In cases where `pErr != nil`, should we be replacing the error or wrapping it?
</blockquote></details>

You are probably right that wrapping is better here. It would be less likely to break any existing tests, however I'm not sure in practice that it should make a difference. Both of these "early termination" errors should be treated similarly to `AmbiguousResultErrors`.
___
*[`pkg/kv/txn_external_test.go` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1IfwxDstr50Jp2I4y:-Nm1Wjo_0a7dPxPZZ832:b8whpa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/txn_external_test.go#L258)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is this change okay? Don't we want to return an `AmbiguousResultError` in this case based on the commentary above?
</blockquote></details>

This is a good question, and I don't know for sure the answer. It still FEELS like it any callers should treat the other errors as ambiguous, but its not clear they all do today. From a "context canceled" perspective, presumably the canceller is  the caller, so I think they won't even look at the error regardless. From a "shutdown" error, my guess is that the new error is handled better and as we see in the broken test, it at least handles this error better.

The tricky case today is a non-terminal error where we stop retrying because of a Context error. I'm not sure we want to wrap that error, and the fact that we would have otherwise retried means that we are in a non-terminal case.

**Comment by andrewbaptist** [comment]
Also - maybe I shouldn't backport this. The issue isn't new, and the only test that caught this problem https://github.com/cockroachdb/cockroach/issues/113125 was a rare race condition. That would allow this to bake on master for a long time.

**Review by andrewbaptist — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116583)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2027 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1Kd7L98zR1LV0iTkD:-NnogDod50b0rTr_6xCi:bkwi19e) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/kvclient/kvcoord/dist_sender.go#L2027)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

You are probably right that wrapping is better here. It would be less likely to break any existing tests, however I'm not sure in practice that it should make a difference. Both of these "early termination" errors should be treated similarly to `AmbiguousResultErrors`.
</blockquote></details>

The change now Joins it which allows easier detection of either type of error. There is no perfect answer here as our error handling above KV is not completely consistent. 
___
*[`pkg/kv/txn_external_test.go` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1IfwxDstr50Jp2I4y:-NnogoKQ7wIUklfrVUhQ:b-p6ozk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/txn_external_test.go#L258)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This is a good question, and I don't know for sure the answer. It still FEELS like it any callers should treat the other errors as ambiguous, but its not clear they all do today. From a "context canceled" perspective, presumably the canceller is  the caller, so I think they won't even look at the error regardless. From a "shutdown" error, my guess is that the new error is handled better and as we see in the broken test, it at least handles this error better.

The tricky case today is a non-terminal error where we stop retrying because of a Context error. I'm not sure we want to wrap that error, and the fact that we would have otherwise retried means that we are in a non-terminal case.
</blockquote></details>

This now returns a joined error so either type is handled.

**Review by arulajmani — COMMENTED** [review]
The error handling here still doesn't feel quite right. See my comment below.

> Reviewed 2 of 8 files at r2, 3 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-:-Np1LPXMB83gtx-ftsaf:b-okayn0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2036 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Np1Buw51FYzgq5W6WgP:-Np1Buw51FYzgq5W6WgQ:bhx4fue) ([raw file](https://github.com/cockroachdb/cockroach/blob/f39378d1fcd0c96fb99b2717ae8cef1926d2399d/pkg/kv/kvclient/kvcoord/dist_sender.go#L2036)):*
> ```Go
> 		// Happens when the client request is canceled or the deadline is
> 		// exceeded. We wrap to have a consistent error to check against.
> 		return errors.Join(err, ctx.Err())
> ```

Let's errors.Wrap(errors.Join(err, ctx.Err()), "aborted in DistSender") to keep with the old behaviour and stay true to the comment above.

EDIT -- probably superseded by my comment below.
___
*[`pkg/kv/kvclient/kvcoord/transport.go` line 201 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Np1C9XI6D8z7ApPIKKh:-Np1C9XI6D8z7ApPIKKi:bmsgii7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f39378d1fcd0c96fb99b2717ae8cef1926d2399d/pkg/kv/kvclient/kvcoord/transport.go#L201)):*
> ```Go
> 	// in the local server comes pretty late)
> 	if err := ctx.Err(); err != nil {
> 		return nil, err
> ```

nit revert?
___
*[`pkg/kv/txn_external_test.go` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1IfwxDstr50Jp2I4y:-Np1CQyy3pJ6cUfhgMrN:b3e7s9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/txn_external_test.go#L258)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This now returns a joined error so either type is handled.
</blockquote></details>

Looks like even with the `Join` we're not able to bring back the `require.IsType(t, &kvpb.AmbiguousResultError{}, commitErr)` line. This means that checks of the form:

```go
if !errors.HasType(commitErr, (\*kvpb.AmbiguousResultError)(nil)) {  
 // Do something ...
 }
```

won't work. We rely on such checks in other places in the code, so I don't think our use of `Join` is giving us what we need here.

We're having to change this and other tests with similar assertions (in spirit). We shouldn't need to do so -- let's revert such changes, and get them passing, before we merge this patch.

**Review by andrewbaptist — COMMENTED** [review]
TFTR - This is tricky to keep the old behavior and handle this well. I think this is a lot better now and shouldn't add any changes above.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-:-Np5FfUB9nk5VTHc7fM0:b46k8ai)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2036 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Np1Buw51FYzgq5W6WgP:-Np5NL9h1NbxQxbGYkOI:b1oih3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f39378d1fcd0c96fb99b2717ae8cef1926d2399d/pkg/kv/kvclient/kvcoord/dist_sender.go#L2036)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's errors.Wrap(errors.Join(err, ctx.Err()), "aborted in DistSender") to keep with the old behaviour and stay true to the comment above.

EDIT -- probably superseded by my comment below.
</blockquote></details>

Yep - I simplified this to convert "all" errors that are due to context cancelation to be Ambiguous. This at least returns an error that clients can handle and might be slightly pessimistic but should be OK for most cases including the original test issue. 
___
*[`pkg/kv/txn_external_test.go` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1IfwxDstr50Jp2I4y:-Np5NjWx81zOmAb0hQ5H:b-77b0d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/txn_external_test.go#L258)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Looks like even with the `Join` we're not able to bring back the `require.IsType(t, &kvpb.AmbiguousResultError{}, commitErr)` line. This means that checks of the form:

```go
if !errors.HasType(commitErr, (\*kvpb.AmbiguousResultError)(nil)) {  
 // Do something ...
 }
```

won't work. We rely on such checks in other places in the code, so I don't think our use of `Join` is giving us what we need here.

We're having to change this and other tests with similar assertions (in spirit). We shouldn't need to do so -- let's revert such changes, and get them passing, before we merge this patch.
</blockquote></details>

Tests all rolled back and pass. 
___
*[`pkg/kv/kvclient/kvcoord/transport.go` line 201 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Np1C9XI6D8z7ApPIKKh:-Np5No2PCgvXOItlYEM7:bct7pp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f39378d1fcd0c96fb99b2717ae8cef1926d2399d/pkg/kv/kvclient/kvcoord/transport.go#L201)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit revert?
</blockquote></details>

reverted. I'm not sure it mattered in practice, but safer to just wrap more times then less :)

**Review by arulajmani — APPROVED** [review]
:lgtm: modulo comments. Thanks for working through the review iterations on this one -- this stuff is tricky, but I think we got it in a good place at the end.

> Reviewed 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-:-Np6IWSf8J62Js6oRSil:berxpdn)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2031 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Np6IA588xkKk2NCDVVu:-Np6IA588xkKk2NCDVVv:bufjy2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/380b48d8c75fb10375b8e1e88480cdb68f721727/pkg/kv/kvclient/kvcoord/dist_sender.go#L2031)):*
> ```Go
> 	select {
> 	case <-ds.rpcRetryOptions.Closer:
> 		return errors.Wrap(kvpb.NewAmbiguousResultError(err), "aborted in DistSender (node unavailable)")
> ```

One way to get around hard coding "context canceled" and "node unavailable" here is to do something like: 

```go
diff --git a/pkg/kv/kvclient/kvcoord/dist_sender.go b/pkg/kv/kvclient/kvcoord/dist_sender.go
index b1633e0f6f8..3473fbdd2fe 100644
--- a/pkg/kv/kvclient/kvcoord/dist_sender.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender.go
@@ -2028,9 +2028,9 @@ func (ds *DistSender) deduceRetryEarlyExitError(ctx context.Context, err error)
        }
        select {
        case <-ds.rpcRetryOptions.Closer:
-               return errors.Wrap(kvpb.NewAmbiguousResultError(err), "aborted in DistSender (node unavailable)")
+               return errors.Wrapf(kvpb.NewAmbiguousResultError(errors.CombineErrors(&kvpb.NodeUnavailableError{}, err)), "aborted in DistSender")
        case <-ctx.Done():
-               return errors.Wrap(kvpb.NewAmbiguousResultError(err), "aborted in DistSender (context canceled)")
+               return errors.Wrapf(kvpb.NewAmbiguousResultError(errors.CombineErrors(ctx.Err(), err)), "aborted in DistSender")
        default:
        }
        return nil
```
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 649 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Np6Fv1F7O8GY8aMvCvG:-Np6Fv1F7O8GY8aMvCvH:bx2ucy4) ([raw file](https://github.com/cockroachdb/cockroach/blob/380b48d8c75fb10375b8e1e88480cdb68f721727/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L649)):*
> ```Go
> 		// TODO(baptist): Currently this doesn't detect if the context is cancelled
> 		// prior to entering the loop the first time. Currently Retry.Next always
> 		// guarantees to run at least one time due to
> ```

Unfinished sentence?
___
*[`pkg/kv/txn_external_test.go` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1IfwxDstr50Jp2I4y:-Np5nI6sAN73wzPDB8Yv:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/txn_external_test.go#L258)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Tests all rolled back and pass.
</blockquote></details>

Nice!

**Review by andrewbaptist — COMMENTED** [review]
Thank you! This was a tricky one to get right.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-:-Np6eJYt7kkmFHQ7srlC:bai1eiq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2031 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Np6IA588xkKk2NCDVVu:-Np6fIFy7Zhot4CZC2vi:bn6vw0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/380b48d8c75fb10375b8e1e88480cdb68f721727/pkg/kv/kvclient/kvcoord/dist_sender.go#L2031)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

One way to get around hard coding "context canceled" and "node unavailable" here is to do something like: 

```go
diff --git a/pkg/kv/kvclient/kvcoord/dist_sender.go b/pkg/kv/kvclient/kvcoord/dist_sender.go
index b1633e0f6f8..3473fbdd2fe 100644
--- a/pkg/kv/kvclient/kvcoord/dist_sender.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender.go
@@ -2028,9 +2028,9 @@ func (ds *DistSender) deduceRetryEarlyExitError(ctx context.Context, err error)
        }
        select {
        case <-ds.rpcRetryOptions.Closer:
-               return errors.Wrap(kvpb.NewAmbiguousResultError(err), "aborted in DistSender (node unavailable)")
+               return errors.Wrapf(kvpb.NewAmbiguousResultError(errors.CombineErrors(&kvpb.NodeUnavailableError{}, err)), "aborted in DistSender")
        case <-ctx.Done():
-               return errors.Wrap(kvpb.NewAmbiguousResultError(err), "aborted in DistSender (context canceled)")
+               return errors.Wrapf(kvpb.NewAmbiguousResultError(errors.CombineErrors(ctx.Err(), err)), "aborted in DistSender")
        default:
        }
        return nil
```
</blockquote></details>

I'm trying this now - hopefully this all works. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 649 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Np6Fv1F7O8GY8aMvCvG:-Np6eTIC-AejKuW4dS2A:b-nffsuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/380b48d8c75fb10375b8e1e88480cdb68f721727/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L649)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Unfinished sentence?
</blockquote></details>

I removed this TODO. I had forgotten that this was the desired behavior. 
___
*[`pkg/kv/txn_external_test.go` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116583#-Nm1IfwxDstr50Jp2I4y:-Np6eXwh2tP9qqY-oh5H:b2nhouv) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca74ab71b602e6e0d6be40308af552e89740dfd/pkg/kv/txn_external_test.go#L258)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Nice!
</blockquote></details>

TY!

**Comment by andrewbaptist** [comment]
bors r=arulajmani

**Comment by andrewbaptist** [comment]
bors retry

**Comment by andrewbaptist** [comment]
bors retry

Hit #117596

**Pull request merged**
